daffodil-lib/src/main/scala/edu/illinois/ncsa/daffodil/util/DecimalUtils.scala

```diff
@@ -0,0 +1,223 @@
+
+import java.math.{ BigInteger => JBigInteger, BigDecimal => JBigDecimal }
+
+object DecimalUtils {
+
+  def packedToBigDecimal(packedNum: Array[Byte], scale: Int): JBigDecimal = {
+    val packed: AS400PackedDecimal = new AS400PackedDecimal((packedNum.length * 2) - 1, scale)
+
+    packed.toObject(packedNum).asInstanceOf[JBigDecimal]
+  }
+
+  def packedFromBigDecimal(num: JBigDecimal, nBits: Int): Array[Byte] = {
+    val packed: AS400PackedDecimal = new AS400PackedDecimal(nBits/4 - 1, num.scale)
+
+    packed.toBytes(num)
+  }
+
+  def bcdToBigInteger(bcdNum: Array[Byte]): JBigInteger = {
+    val numDigits: Int = bcdNum.size * 2  // 2 digits stored per byte
+    val outputData: Array[Char] = new Array[Char](numDigits)
+    var outputPos: Int = 0
+    var offset: Int = 0
+    var nibble: Int = 0
+
+    while (offset < bcdNum.size) {
+      nibble = (bcdNum(offset) & 0xFF) >>> 4
+      if (nibble > 0x09)
+        throw new NumberFormatException("Invalid high nibble")
+
+      outputData(outputPos) = (nibble | 0x0030).toChar
+      outputPos = outputPos + 1
+
+      nibble = (bcdNum(offset) & 0x0F)
+      offset = offset + 1
+      if (nibble > 0x09)
+        throw new NumberFormatException("Invalid low nibble")
+
+      outputData(outputPos) = (nibble | 0x0030).toChar
+      outputPos = outputPos + 1
+    }
+
+    new JBigInteger(new String(outputData))
+  }
+
+  def bcdToBigDecimal(bcdNum: Array[Byte], scale: Int): JBigDecimal = {
+    new JBigDecimal(bcdToBigInteger(bcdNum), scale)
+  }
+
+  def bcdFromBigInteger(num: JBigInteger, nBits: Int): Array[Byte] = {
+    val inChars: Array[Char] = num.toString.toCharArray
+    var offset: Int = 0
+    var inPos: Int = 0
+    var firstNibble: Int = 0
+    var secondNibble: Int = 0
+    val outArray: Array[Byte] = new Array[Byte](inChars.length / 2 + 1)
+    val leadingZeros: Int = if(inChars.length % 2 == 0) (nBits/4 - inChars.length + 1) else (nBits/4 - inChars.length)
+
+    // Add leading zeros if necessary
+    while(offset < (leadingZeros - 1)) {
+      outArray(offset) = 0
+      offset = offset + 1
+    }
+
+    // Need even number of digits, pad with 0 if necessary
+    if (inChars.length % 2 != 0) {
+      outArray(offset) = (inChars(inPos) & 0x000F).asInstanceOf[Byte]
+      offset = offset + 1
+      inPos = inPos + 1
+    }
+
+    while (inPos < inChars.length) {
+      firstNibble = (inChars(inPos) & 0x000F) << 4
+      inPos = inPos + 1
+      secondNibble = inChars(inPos) & 0x000F
+      inPos = inPos + 1
+      outArray(offset) = (firstNibble + secondNibble).asInstanceOf[Byte]
+      offset = offset + 1
+    }
+
+    outArray
+  }
+
+  def bcdFromBigDecimal(num: JBigDecimal, nBits: Int): Array[Byte] = {
+    bcdFromBigInteger(num.unscaledValue, nBits)
+  }
+
+  def IBM4690ToBigInteger(num: Array[Byte]): JBigInteger = {
+    val numDigits: Int = num.size * 2  // 2 digits stored per byte
+    val outputData: Array[Char] = new Array[Char](numDigits)
```

Comment:
Same as above, might be more efficient to just build up the number rather than building an char array and converting to a number later.